File: order.service/src/test/java/br/com/edu/order/domain/OrderTest.java
package br.com.edu.order.domain;

import br.com.edu.order.domain.errors.InvalidPriceException;
import br.com.edu.order.domain.errors.InvalidQuantityException;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;

import static org.junit.jupiter.api.Assertions.*;

class OrderTest {

    @Test
    @DisplayName("Should create with default values")
    void shouldCreateDefaultValues() {
        final var order = new Order();
        assertEquals(Status.PENDING, order.getStatus());
        assertNotNull(order.getCreated());
        assertNotNull(order.getUpdated());
        assertNotNull(order.getId());
        assertEquals(BigDecimal.ZERO, order.getTotal());
    }

    @Test
    @DisplayName("Should calculate the total value")
    void shouldCalculateTotal() {

        final var item = new Item();
        item.setQuantity(2);
        item.setProduct("test product");
        item.setPrice(BigDecimal.valueOf(0.01));

        final var order = new Order();
        order.addItem(item);

        var total = order.getTotal();
        assertEquals(BigDecimal.valueOf(0.02), total);
    }

    @Test
    @DisplayName("Should throw InvalidPriceException when the price is zero")
    void shouldThrowInvalidPriceWithPriceZero() {
        assertThrows(InvalidPriceException.class, () -> {
            final var item = new Item();
            item.setQuantity(2);
            item.setProduct("test product");
            item.setPrice(BigDecimal.valueOf(0.00));
        });
    }

    @Test
    @DisplayName("Should throw InvalidPriceException when the price is negative")
    void shouldThrowInvalidPriceWithNegative() {
        assertThrows(InvalidPriceException.class, () -> {
            final var item = new Item();
            item.setQuantity(2);
            item.setProduct("test product");
            item.setPrice(BigDecimal.valueOf(-0.01));
        });
    }

    @Test
    @DisplayName("Should throw InvalidQuantity when the quantity is zero")
    void shouldThrowInvalidQuantityWithPriceZero() {
        assertThrows(InvalidQuantityException.class, () -> {
            final var item = new Item();
            item.setQuantity(0);
            item.setProduct("test product");
            item.setPrice(BigDecimal.valueOf(0.01));
        });
    }

    @Test
    @DisplayName("Should throw InvalidQuantity when the quantity is negative")
    void shouldThrowInvalidQuantityWithNegative() {
        assertThrows(InvalidQuantityException.class, () -> {
            final var item = new Item();
            item.setQuantity(-1);
            item.setProduct("test product");
            item.setPrice(BigDecimal.valueOf(0.01));
        });
    }

}

File: order.service/order-service-test/src/test/kotlin/order/domain/OrderTest.kt
package order.domain

import order.domain.errors.InvalidPriceException
import order.domain.errors.InvalidQuantityException
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import java.math.BigDecimal
import java.math.RoundingMode

internal class OrderTest {

    @Test
    @DisplayName("Should create with default values")
    fun shouldCreateDefaultValues() {
        val order = Order()
        assertEquals(Status.PENDING, order.status)
        assertNotNull(order.created)
        assertNotNull(order.updated)
        assertNotNull(order.id)
        assertEquals(BigDecimal.ZERO.setScale(2, RoundingMode.HALF_UP), order.total)
    }

    @Test
    @DisplayName("Should calculate the total value")
    fun shouldCalculateTotal() {
        val item = Item()
        item.quantity = 2
        item.product = "test product"
        item.price = BigDecimal.valueOf(0.01)

        val order = Order()
        order.addItem(item)

        val total = order.total
        assertEquals(BigDecimal.valueOf(0.02), total)
    }

    @Test
    @DisplayName("Should throw InvalidPriceException when the price is zero")
    fun shouldThrowInvalidPriceWithPriceZero() {
        assertThrows(InvalidPriceException::class.java) {
            val item = Item()
            item.quantity = 2
            item.product = "test product"
            item.price = BigDecimal.valueOf(0.00)
        }
    }

    @Test
    @DisplayName("Should throw InvalidPriceException when the price is negative")
    fun shouldThrowInvalidPriceWithNegative() {
        assertThrows(InvalidPriceException::class.java) {
            val item = Item()
            item.quantity = 2
            item.product = "test product"
            item.price = BigDecimal.valueOf(-0.01)
        }
    }

    @Test
    @DisplayName("Should throw InvalidQuantity when the quantity is zero")
    fun shouldThrowInvalidQuantityWithPriceZero() {
        assertThrows(InvalidQuantityException::class.java) {
            val item = Item()
            item.quantity = 0
            item.product = "test product"
            item.price = BigDecimal.valueOf(0.01)
        }
    }

    @Test
    @DisplayName("Should throw InvalidQuantity when the quantity is negative")
    fun shouldThrowInvalidQuantityWithNegative() {
        assertThrows(InvalidQuantityException::class.java) {
            val item = Item()
            item.quantity = -1
            item.product = "test product"
            item.price = BigDecimal.valueOf(0.01)
        }
    }

}
